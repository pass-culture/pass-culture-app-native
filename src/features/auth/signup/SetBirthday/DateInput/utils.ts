export interface DateValidation {
  isComplete: boolean
  isValid: boolean
  isDateAboveMin: boolean
  isDateBelowMax: boolean
}

export interface DateInputRef {
  clearFocuses: () => void
}
