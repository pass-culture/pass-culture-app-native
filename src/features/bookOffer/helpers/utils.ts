import { OfferStockResponse } from 'api/gen'

export enum OfferStatus {
  BOOKABLE = 'BOOKABLE',
  NOT_BOOKABLE = 'NOT_BOOKABLE',
  NOT_OFFERED = 'NOT_OFFERED',
}

export const formatToKeyDate = (dateString: Date | string) => {
  const date = new Date(dateString)
  const day = ('0' + date.getDate()).slice(-2)
  const month = ('0' + (date.getMonth() + 1)).slice(-2)
  const year = date.getFullYear()
  return `${year}-${month}-${day}`
}

export const formatHour = (dateString: string | null | undefined): string => {
  if (!dateString) return ''
  const date = new Date(dateString)
  const hours = date.getHours()
  const minutes = ('0' + date.getMinutes()).slice(-2)
  return `${hours}:${minutes}`
}

export const getStatusFromStocks = (stocks: OfferStockResponse[], credit: number): OfferStatus => {
  const bookableStocks = stocks.filter(({ isBookable }) => isBookable)
  if (bookableStocks.length === 0) return OfferStatus.NOT_BOOKABLE
  return bookableStocks.some(({ price }) => price <= credit)
    ? OfferStatus.BOOKABLE
    : OfferStatus.NOT_BOOKABLE
}
