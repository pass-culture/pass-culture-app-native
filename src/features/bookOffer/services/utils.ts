import { OfferStockResponse } from 'api/gen'
import { formatToSlashedFrenchDate } from 'libs/dates'
import { CENTS_IN_EURO } from 'libs/parsers/pricesConversion'

const EURO_SYMBOL = '€'

export const getStocksByDate = (
  stocks: OfferStockResponse[]
): { [date: string]: OfferStockResponse[] } => {
  const stockDates: { [date: string]: OfferStockResponse[] } = {}

  stocks.forEach((stock) => {
    if (stock.beginningDatetime !== null && stock.beginningDatetime !== undefined) {
      const formattedDate = formatToSlashedFrenchDate(stock.beginningDatetime.toString())
      stockDates[formattedDate] = stockDates[formattedDate]
        ? stockDates[formattedDate].concat(stock)
        : [stock]
    }
  })

  return stockDates
}

export enum OfferStatus {
  BOOKABLE = 'BOOKABLE',
  NOT_BOOKABLE = 'NOT_BOOKABLE',
  NOT_OFFERED = 'NOT_OFFERED',
}

export const formatToFrenchDecimalWithoutSpace = (cents: number) => {
  const euros = cents / CENTS_IN_EURO
  // we show 2 decimals if price is not round. Ex: 21,50€
  const fixed = euros === Math.floor(euros) ? euros : euros.toFixed(2)
  return `${fixed.toString().replace('.', ',')}${EURO_SYMBOL}`
}

export const getDateStatusAndPrice = (
  date: Date,
  stocksDates: { [date: string]: OfferStockResponse[] }
): { status: OfferStatus; price: string | null } => {
  const stocksByDate = stocksDates[formatToSlashedFrenchDate(date.toString())]
  if (!stocksByDate) return { status: OfferStatus.NOT_OFFERED, price: null }

  const prices = stocksByDate.map((stock) => stock.price)
  const price = formatToFrenchDecimalWithoutSpace(Math.min(...prices))
  const offerIsBookable = stocksByDate.some((stock) => stock.isBookable)

  if (offerIsBookable) return { status: OfferStatus.BOOKABLE, price }
  return { status: OfferStatus.NOT_BOOKABLE, price }
}
