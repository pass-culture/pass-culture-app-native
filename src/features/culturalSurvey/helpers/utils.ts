import {
  CulturalSurveyQuestionEnum,
  CulturalSurveyQuestionsResponse,
  UserProfileResponse,
} from 'api/gen/api'
import { useSettingsContext } from 'features/auth/SettingsContext'

export const mapQuestionIdToPageTitle = (id: CulturalSurveyQuestionEnum | undefined) => {
  switch (id) {
    case CulturalSurveyQuestionEnum.SORTIES:
    case CulturalSurveyQuestionEnum.FESTIVALS:
    case CulturalSurveyQuestionEnum.SPECTACLES:
      return 'Tes sorties'
    case CulturalSurveyQuestionEnum.ACTIVITES:
      return 'Tes activitÃ©s'
    default:
      return ''
  }
}

export function useCulturalSurveyRoute() {
  const { data: settings } = useSettingsContext()

  // as long as ENABLE_CULTURAL_SURVEY is still used in the api to define if user should
  // fill cultural survey, we use enableNativeCulturalSurvey to know which cultural survey
  // (native or typeform) to show
  if (settings?.enableNativeCulturalSurvey) {
    return 'CulturalSurveyIntro'
  }
  return 'CulturalSurvey'
}

export function shouldShowCulturalSurvey(user?: UserProfileResponse) {
  if (!user) return false
  return user.needsToFillCulturalSurvey
}

export const createInitialQuestionsList = (
  culturalSurveyQuestions?: CulturalSurveyQuestionsResponse
) => {
  const subQuestions = [] as CulturalSurveyQuestionEnum[]
  const initialQuestionsList = [] as CulturalSurveyQuestionEnum[]

  if (!culturalSurveyQuestions?.questions) return initialQuestionsList

  culturalSurveyQuestions.questions.forEach((question) => {
    if (!subQuestions.includes(question.id)) {
      initialQuestionsList.push(question.id)
    }
    question.answers.forEach((answer) => {
      if (answer.sub_question && !subQuestions.includes(answer.sub_question)) {
        subQuestions.push(answer.sub_question)
      }
    })
  })
  return initialQuestionsList
}

export const addSubQuestionToQuestionsToDisplay = (
  subQuestionId: CulturalSurveyQuestionEnum,
  questionsToDisplay: CulturalSurveyQuestionEnum[]
) => {
  let updatedQuestionsToDisplay = [...questionsToDisplay]
  updatedQuestionsToDisplay = updatedQuestionsToDisplay.filter(
    (questionId) => questionId != subQuestionId
  )
  return updatedQuestionsToDisplay
}

export const removeSubQuestionsToDisplay = (
  subQuestionId: CulturalSurveyQuestionEnum,
  questionsToDisplay: CulturalSurveyQuestionEnum[]
) => {
  const updatedQuestionsToDisplay = [...questionsToDisplay]
  updatedQuestionsToDisplay.splice(1, 0, subQuestionId)
  return updatedQuestionsToDisplay
}
