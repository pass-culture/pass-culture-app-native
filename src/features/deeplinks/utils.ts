import { env } from 'libs/environment'

export const FIREBASE_DYNAMIC_LINK_URL = `https://${env.FIREBASE_DYNAMIC_LINK_DOMAIN}`

/**
 * @see https://firebase.google.com/docs/dynamic-links/create-manually
 */
export function getLongDynamicLinkURI() {
  return `apn=${env.ANDROID_APP_ID}&isi=${env.IOS_APP_STORE_ID}&ibi=${env.IOS_APP_ID}&efr=1`
}

/**
 * @see https://firebase.google.com/docs/dynamic-links/create-manually
 * @param deepLink The deeplink targeted screen
 * @param deepLinkParams Firebase additional params https://firebase.google.com/docs/dynamic-links/create-manually#parameters or ofl URL if a string. TODO: remove string support once webapp migration is complete
 */
export function generateLongFirebaseDynamicLink(
  deepLink: string,
  deepLinkParams?: Record<string, string>
) {
  let params = ''
  if (deepLinkParams) {
    Object.entries(deepLinkParams).forEach(([key, value]) => {
      params += `&${key}=${value}`
    })
  }
  return `${FIREBASE_DYNAMIC_LINK_URL}/?link=${encodeURIComponent(
    deepLink
  )}&${getLongDynamicLinkURI()}${params}`
}

export const isFirebaseDynamicLink = (url: string) => url.startsWith(FIREBASE_DYNAMIC_LINK_URL)
export const isFirebaseLongDynamicLink = (url: string) =>
  isFirebaseDynamicLink(url) && url.includes('?link=')
