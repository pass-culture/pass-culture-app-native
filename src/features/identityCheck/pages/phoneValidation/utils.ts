export const formatPhoneNumberWithPrefix = (phoneNumber: string, countryCallingCode: string) => {
  let preformattedNumber = phoneNumber.replace(/[\s.-]*/g, '')
  if (preformattedNumber.startsWith('0')) {
    preformattedNumber = preformattedNumber.substring(1)
  }
  return `+${countryCallingCode}${preformattedNumber}`
}
