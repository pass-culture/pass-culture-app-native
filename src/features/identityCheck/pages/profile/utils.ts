import {
  ActivityIdEnum,
  ActivityResponseModel,
  SchoolTypeResponseModel,
  SchoolTypesIdEnum,
} from 'api/gen'

export const mapSchoolTypeIdToLabelAndDescription = (
  schoolTypeId: SchoolTypesIdEnum,
  schoolTypes: SchoolTypeResponseModel[]
) => {
  const schoolType = schoolTypes?.find((schoolType) => schoolType.id === schoolTypeId)
  return { label: schoolType?.label, description: schoolType?.description }
}

export const getSchoolTypesIdsFromActivity = (
  activityId: ActivityIdEnum,
  activities: ActivityResponseModel[]
) => {
  const activitySchoolTypes = activities.find(
    (activity) => activity.id === activityId
  )?.associatedSchoolTypesIds

  return activitySchoolTypes
}

export const activityHasSchoolTypes = (
  activityId: ActivityIdEnum,
  activities: ActivityResponseModel[]
) => {
  return !!getSchoolTypesIdsFromActivity(activityId, activities)?.length
}
