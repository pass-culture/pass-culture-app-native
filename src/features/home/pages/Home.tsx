import { t } from '@lingui/macro'
import { useFocusEffect, useNavigation, useRoute } from '@react-navigation/native'
import React, { FunctionComponent, useState } from 'react'
import { NativeSyntheticEvent, NativeScrollEvent, ScrollView, TouchableOpacity } from 'react-native'
import { getStatusBarHeight } from 'react-native-iphone-x-helper'
import styled from 'styled-components/native'

import { useUserProfileInfo } from 'features/home/api'
import { HomeBodyPlaceholder, SignUpSignInChoiceModal } from 'features/home/components'
import { HomeBody } from 'features/home/components/HomeBody'
import { useDisplayedHomeModules } from 'features/home/pages/useDisplayedHomeModules'
import { useAvailableCredit } from 'features/home/services/useAvailableCredit'
import { UseNavigationType, UseRouteType } from 'features/navigation/RootNavigator'
import { analytics } from 'libs/analytics'
import { isCloseToBottom } from 'libs/analytics.utils'
import { env } from 'libs/environment'
import { _ } from 'libs/i18n'
import { formatToFrenchDecimal } from 'libs/parsers'
import { useModal } from 'ui/components/modals/useModal'
import { HeaderBackground } from 'ui/svg/HeaderBackground'
import { UserCircle } from 'ui/svg/icons/UserCircle'
import { ColorsEnum, getSpacing, Spacer, Typo } from 'ui/theme'
import { ACTIVE_OPACITY } from 'ui/theme/colors'
import { ZIndexes } from 'ui/theme/layers'

import { useShowSkeleton } from './useShowSkeleton'

const statusBarHeight = getStatusBarHeight(true)

export const Home: FunctionComponent = function () {
  const navigation = useNavigation<UseNavigationType>()
  const { params } = useRoute<UseRouteType<'Home'>>()
  const { data: userInfos } = useUserProfileInfo()
  const { visible: signInModalVisible, showModal: showSignInModal, hideModal } = useModal(false)
  const showSkeleton = useShowSkeleton()
  const availableCredit = useAvailableCredit()

  const [hasSeenAllModules, setHasSeenAllModules] = useState<boolean>(false)
  const { displayedModules, algoliaModules } = useDisplayedHomeModules()

  function hideSignInModal() {
    navigation.setParams({ shouldDisplayLoginModal: false })
    hideModal()
  }

  useFocusEffect(() => {
    if (params.shouldDisplayLoginModal) {
      showSignInModal()
    }
  })

  const checkIfAllModulesHaveBeenSeen = ({
    nativeEvent,
  }: NativeSyntheticEvent<NativeScrollEvent>) => {
    if (!hasSeenAllModules && isCloseToBottom(nativeEvent)) {
      setHasSeenAllModules(true)
      analytics.logAllModulesSeen(displayedModules.length)
    }
  }

  let subtitle = _(t`Toute la culture dans ta main`)
  if (availableCredit) {
    subtitle = availableCredit.isExpired
      ? _(t`Ton crédit est expiré`)
      : _(t`Tu as ${formatToFrenchDecimal(availableCredit.amount)} sur ton pass`)
  }

  return (
    <ScrollView
      testID="homeScrollView"
      scrollEventThrottle={400}
      bounces={false}
      onScroll={checkIfAllModulesHaveBeenSeen}>
      <Spacer.TopScreen />
      {env.CHEAT_BUTTONS_ENABLED && (
        <CheatButtonsContainer>
          <CheatTouchableOpacity onPress={() => navigation.navigate('AppComponents')}>
            <Typo.Body>{_(t`Composants`)}</Typo.Body>
          </CheatTouchableOpacity>
          <CheatTouchableOpacity onPress={() => navigation.navigate('Navigation')}>
            <Typo.Body>{_(t`Navigation`)}</Typo.Body>
          </CheatTouchableOpacity>
        </CheatButtonsContainer>
      )}
      <HeaderBackgroundWrapper>
        <HeaderBackground />
      </HeaderBackgroundWrapper>
      <UserProfileContainer onPress={() => navigation.navigate('CheatMenu')}>
        <UserCircle size={32} color={ColorsEnum.WHITE} />
      </UserProfileContainer>

      <CenterContainer>
        <Spacer.Column numberOfSpaces={8} />
        <StyledTitle1 color={ColorsEnum.WHITE} numberOfLines={2}>
          {userInfos?.firstName
            ? _(/*i18n: Hello title message */ t`Bonjour ${userInfos?.firstName}`)
            : _(/*i18n: Welcome title message */ t`Bienvenue !`)}
        </StyledTitle1>
        <Spacer.Column numberOfSpaces={2} />
        <Typo.Body color={ColorsEnum.WHITE}>{subtitle}</Typo.Body>
      </CenterContainer>
      <Spacer.Column numberOfSpaces={6} />

      {showSkeleton ? <HomeBodyPlaceholder /> : null}
      <HomeBodyLoadingContainer isLoading={showSkeleton}>
        <HomeBody modules={displayedModules} algoliaModules={algoliaModules} />
      </HomeBodyLoadingContainer>

      <SignUpSignInChoiceModal visible={signInModalVisible} dismissModal={hideSignInModal} />
      <Spacer.TabBar />
    </ScrollView>
  )
}

const StyledTitle1 = styled(Typo.Title1)({
  textAlign: 'center',
  marginHorizontal: getSpacing(8),
})

const CenterContainer = styled.View({
  flex: 1,
  alignItems: 'center',
})

const HeaderBackgroundWrapper = styled.View({
  position: 'absolute',
  top: 0,
  left: 0,
  zIndex: 0,
})

const HomeBodyLoadingContainer = styled.View<{ isLoading: boolean }>(({ isLoading }) => ({
  height: isLoading ? 0 : undefined,
  overflow: 'hidden',
}))

const CheatButtonsContainer = styled.View({
  width: '100%',
  flexDirection: 'row',
  justifyContent: 'space-around',
  zIndex: ZIndexes.HomeCheatButton,
  top: 20,
})

const UserProfileContainer = styled.TouchableOpacity.attrs({
  activeOpacity: ACTIVE_OPACITY,
})({
  position: 'absolute',
  right: 24,
  top: getSpacing(3) + statusBarHeight,
  zIndex: 1,
})

const CheatTouchableOpacity = styled(TouchableOpacity)({
  borderColor: ColorsEnum.BLACK,
  borderWidth: 2,
  padding: getSpacing(1),
})
