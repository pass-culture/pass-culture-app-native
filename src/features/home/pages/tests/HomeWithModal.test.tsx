import { NavigationContainer } from '@react-navigation/native'
import { render, act } from '@testing-library/react-native'
import React from 'react'

import { Tab } from 'features/navigation/TabBar/TabNavigator'
import { reactQueryProviderHOC } from 'tests/reactQueryProviderHOC'
import { flushAllPromises } from 'tests/utils'

import { Home } from '../Home'

jest.mock('features/home/api', () => ({
  useUserProfileInfo: jest.fn(() => ({ data: undefined })),
}))

jest.mock('@react-navigation/native', () => jest.requireActual('@react-navigation/native'))
jest.mock('libs/environment', () => ({
  env: {
    FEATURE_FLAG_CODE_PUSH: true,
    CHEAT_BUTTONS_ENABLED: false,
  },
}))

jest.mock('features/auth/AuthContext', () => ({
  useAuthContext: jest.fn(() => ({ isLoggedIn: false })),
}))

jest.mock('features/home/pages/useShowSkeleton', () => ({
  useShowSkeleton: jest.fn(() => false),
}))

jest.mock('features/home/pages/useDisplayedHomeModules', () => ({
  useDisplayedHomeModules: jest.fn(() => ({
    displayedModules: [],
    algoliaModules: {},
  })),
}))

/**
 * We test the modal separately which require
 * the real navigation context to test useFocusEffect
 */
describe('Home component modal test', () => {
  it('should render modal correctly', async () => {
    const homeWithLoginModal = await homeRenderer(true)
    const homeWithoutLoginModal = await homeRenderer(false)

    expect(homeWithoutLoginModal).toMatchDiffSnapshot(homeWithLoginModal)
    homeWithLoginModal.unmount()
    homeWithoutLoginModal.unmount()
  })

  it('should render modal correctly', async () => {
    const { getByTestId, unmount } = await homeRenderer(true)

    expect(getByTestId('modal').props.visible).toBeTruthy()
    unmount()
  })

  it('should not render modal correctly', async () => {
    const { getByTestId, unmount } = await homeRenderer(false)

    expect(getByTestId('modal').props.visible).toBeFalsy()
    unmount()
  })
})

async function homeRenderer(withModal: boolean) {
  const renderAPI = render(
    reactQueryProviderHOC(
      <NavigationContainer>
        <Tab.Navigator initialRouteName="Home">
          <Tab.Screen
            name="Home"
            component={Home}
            initialParams={{ shouldDisplayLoginModal: withModal }}
          />
        </Tab.Navigator>
      </NavigationContainer>
    )
  )
  await act(async () => {
    await flushAllPromises()
  })
  return renderAPI
}
