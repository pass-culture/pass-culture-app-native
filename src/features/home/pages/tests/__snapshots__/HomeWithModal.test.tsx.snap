// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component modal test should render modal correctly 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -220,11 +220,11 @@
                    animationType=\\"slide\\"
                    hardwareAccelerated={false}
                    statusBarTranslucent={true}
                    testID=\\"modal\\"
                    transparent={true}
-                   visible={false}
+                   visible={true}
                  >
                    <View
                      accessible={true}
                      focusable={true}
                      onClick={[Function onClick]}"
`;
